osu.Game/Overlays/RankingsOverlay.cs

```diff
@@ -0,0 +1,198 @@
+
+            background.Colour = OsuColour.Gray(0.1f);
+        }
+
+        protected override void LoadComplete()
+        {
+            country.BindValueChanged(_ => redraw(), true);
+            scope.BindValueChanged(_ => redraw(), true);
+            ruleset.BindValueChanged(_ => redraw(), true);
+            base.LoadComplete();
+        }
+
+        public void ShowCountry(Country requested)
+        {
+            if (requested == null)
+                return;
+
+            Show();
+
+            if (country.Value?.FlagName == requested.FlagName)
+                return;
+
+            country.Value = requested;
+        }
+
+        private void redraw()
+        {
+            scrollFlow.ScrollToStart();
+
+            loading.Show();
+
+            cancellationToken?.Cancel();
+            request?.Cancel();
+
+            cancellationToken = new CancellationTokenSource();
+
+            switch (scope.Value)
+            {
+                default:
+                    contentPlaceholder.Clear();
+                    loading.Hide();
+                    return;
+
+                case RankingsScope.Performance:
+                    createPerformanceTable();
+                    return;
+
+                case RankingsScope.Country:
+                    createCountryTable();
+                    return;
+
+                case RankingsScope.Score:
+                    createScoreTable();
+                    return;
+            }
+        }
+
+        private void createCountryTable()
+        {
+            request = new GetCountryRankingsRequest(ruleset.Value);
+            ((GetCountryRankingsRequest)request).Success += rankings => Schedule(() =>
+            {
+                var table = new CountriesTable(1, rankings.Countries);
+                loadTable(table);
+            });
+
+            api.Queue(request);
+        }
+
+        private void createPerformanceTable()
+        {
+            request = new GetUserRankingsRequest(ruleset.Value, country: country.Value?.FlagName);
+            ((GetUserRankingsRequest)request).Success += rankings => Schedule(() =>
+            {
+                var table = new PerformanceTable(1, rankings.Users);
+                loadTable(table);
+            });
+
+            api.Queue(request);
+        }
+
+        private void createScoreTable()
+        {
+            request = new GetUserRankingsRequest(ruleset.Value, UserRankingsType.Score);
+            ((GetUserRankingsRequest)request).Success += rankings => Schedule(() =>
+            {
+                var table = new ScoresTable(1, rankings.Users);
+                loadTable(table);
+            });
+
+            api.Queue(request);
+        }
+
+        private void loadTable(Drawable table)
+        {
+            LoadComponentAsync(table, t =>
+            {
+                contentPlaceholder.Clear();
+                contentPlaceholder.Add(t);
+                loading.Hide();
+            }, cancellationToken.Token);
```

Comment:
I'm not 100% certain this cancellation logic is bug-free due to the shared field. I'll have to look in detail in an IDE, but if `this` is captured as a closure instead of just `cancellationToken` this might result in races.

Reply:
I have almost the same implementation in `RankingsTablesTestScene`, but if it can be improved - I'm happy to do so